### PR TITLE
Feat/traineedashboard 2.1-2.12-2.4

### PR DESCRIPTION
**Pull request by aplogman**

### CODE REVIEW

File: src/main/java/com/yash/ota/service/TestService.java [hunk 1]
```diff
@@ -102,6 +102,15 @@
 	 */
 	public Test getTestById(int testId) throws InvalidIdException;
 
+	/**
+	 * returns a list of active tests that the user has not completed by their user id
+	 * @param userId int the user id
+	 * @return List/<Test/> the list of tests
+	 * @throws InvalidUserIdException if userId <= 0
```
Comment (neel136846) [inline src/main/java/com/yash/ota/service/TestService.java]: InvalidUserIdException should be InvalidIdException

File: src/main/java/com/yash/ota/serviceimpl/TechnologyServiceImpl.java [hunk 2]
```diff
@@ -84,4 +85,10 @@ public void setDAO(TechnologyDAO mockedDAO) {
 	public List<Technology> getTechnologies() {
 		return technologyDAO.findAll();
 	}
+
```
Comment (neel136846) [inline src/main/java/com/yash/ota/serviceimpl/TechnologyServiceImpl.java]: Add authorship/documentation

File: src/main/java/com/yash/ota/serviceimpl/TestServiceImpl.java [hunk 3]
```diff
@@ -167,6 +167,29 @@ public void setDao(TestDAO dao) {
 	public List<Test> listAll() {
 		return testDAO.findAll();
 	}
+	
+	/**
+	 * @author caleb.reiter
+	 */
+	@Override
+	public List<Test> getAvailableTestsByUserId(int userId) throws InvalidIdException {
```
Comment (neel136846) [inline src/main/java/com/yash/ota/serviceimpl/TestServiceImpl.java]: changed to getAvailableTestByUserId

File: src/main/java/com/yash/ota/exception/InvalidTestIdException.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+/*
+ * InvalidTestIdException
```
Comment (neel136846) [inline src/main/java/com/yash/ota/exception/InvalidTestIdException.java]: Delete this file

File: pom.xml [hunk 5]
```diff
@@ -124,6 +124,16 @@
 				<directory>src/main/java/</directory>
```
Comment (neel136846) [inline pom.xml]: Don't change the pom.xml

Comment (aplogman) [reply pom.xml]: fixed!

File: src/main/java/com/yash/ota/exception/InvalidTestIdException.java (hunk 4 above)
Comment (aplogman) [reply src/main/java/com/yash/ota/exception/InvalidTestIdException.java]: done!

File: src/main/java/com/yash/ota/service/TestService.java (hunk 1 above)
Comment (aplogman) [reply src/main/java/com/yash/ota/service/TestService.java]: done!

File: src/main/java/com/yash/ota/serviceimpl/TechnologyServiceImpl.java (hunk 2 above)
Comment (aplogman) [reply src/main/java/com/yash/ota/serviceimpl/TechnologyServiceImpl.java]: already added in technologyservice class

File: src/main/java/com/yash/ota/serviceimpl/TestServiceImpl.java (hunk 3 above)
Comment (aplogman) [reply src/main/java/com/yash/ota/serviceimpl/TestServiceImpl.java]: fixed!

File: src/test/java/com/yash/ota/serviceimpl/TestServiceImplTest.java [hunk 6]
```diff
@@ -74,6 +74,14 @@ public void testGetAvailableTestsByUserId_InvalidUserIdGiven_ShouldThrow() throw
 		List<com.yash.ota.domain.Test> tests = testService.getAssignedTestsByUserId(-101);
 	}
 
+	
+	/**
+	 * Tests findById method of test service
+	 * 
+	 * @author Zachary Karamanlis
+	 * @throws InvalidIdException 
+	 * @throws InvalidTestIdException
```
[the remark this answers is not in the document]
Comment (aplogman) [reply src/test/java/com/yash/ota/serviceimpl/TestServiceImplTest.java]: changed!

**Pull request merged**